client/src/components/Graph/Graph.tsx

```diff
@@ -1090,24 +1114,115 @@ class Graph extends React.Component<GraphProps, GraphState> {
     prevAsyncProps: GraphAsyncProps | null
   ): void {
     const {
-      positions,
+      positions: positionsEnd,
       colors,
       flags,
       height,
       width,
       imageUnderlay,
       screenCap,
+      layoutChoice,
+      layoutDf,
     } = asyncProps;
-
+    const positionsStart = prevAsyncProps?.positions ?? positionsEnd;
+    const prevLayoutChoice = prevAsyncProps?.layoutChoice;
+    this.duration =
+      layoutChoice.current !== prevLayoutChoice?.current &&
+      prevAsyncProps?.positions
+        ? globals.animationLength
+        : 0;
     this.cachedAsyncProps = asyncProps;
-    const { pointBuffer, colorBuffer, flagBuffer } = this.state;
+
+    const {
+      pointBufferStart,
+      pointBufferEnd,
+      pointBufferTerminal,
+      colorBuffer,
+      flagBuffer,
+    } = this.state;
     let needToRenderCanvas = false;
+
     if (height !== prevAsyncProps?.height || width !== prevAsyncProps?.width) {
       needToRenderCanvas = true;
     }
-    if (positions !== prevAsyncProps?.positions) {
+    if (positionsEnd !== prevAsyncProps?.positions) {
```

Comment:
Done! I've added detailed documentation explaining all four cases of position mapping, the purpose of each buffer, and the logic flow. The documentation covers why different index scenarios require different mapping strategies.

Reply:
I've documented why this is kept inline rather than extracted. The logic is tightly coupled to component state (`this.duration`, `this.state` buffers), needs direct access to `layoutDf.rowIndex`, and the buffer updates are part of the component's render cycle. If you'd prefer it extracted anyway, I can refactor it in a follow-up.